### PR TITLE
Backend/517/admin mfa

### LINKED ISSUE #517: As a super user, I would like my access to the admin dashboard protected by two-factor authentication
**Description:**

From the RFQ:

_The software must have multi-factor authentication for the grantees and PIV authorization for the OFA staff and key personnel. The Login.gov system has already been procured to fulfill this requirement._

https://github.com/18F/tdrs-app-rfq/blob/main/Final-RFQ/FINAL-TDRS-software-development-RFQ.md#32-quality-assurance-surveillance-plan-qasp

**Notes:**

+ This could be accomplished by putting the Django Admin page behind the same Login.gov two-factor flow as the rest of the app. Alternatively, the app could have separate authentication flows for the frontend app and the Django Admin page, both with separate two factor authentication implementation. 

**Acceptance Criteria:** 

- [ ] Django Admin page protected by two factor authentication
- [ ] Leverage login.gov MFA for authentication to the admin site
- [ ] Only users with `staff` can open the admin console
- [ ] Only users with `superuser` can read/update data in the admin console
- [ ] The Admin login page redirects to login.gov

Tasks:
 - [ ] Add BE login url to the settings
 - [ ] Force authentication to the admin login page (which will redirect it to the correct login)

### PR DESCRIPTION
**Pull request by carltonsmith**
### This pull request changes...

Redirects the login to the admin portal to login.gov
Closes #517 

## FUNCTIONALITY
When the user is not logged in to the system and attempts to access the Admin at `/admin/` they will be redirected to the usual login flow to login.gov. Once authenticated, when they navigate to `/admin/` they will see the admin console if they have the proper permissions.

## TO TEST

1. Contact @carltonsmith to get `staff` and `superuser` access
2. Go to https://tdp-frontend.app.cloud.gov/ and ensure that you are _not_ logged in
3. Go to https://tdp-backend.app.cloud.gov/admin/ 
4. If you are redirected to login.gov the change is working correctly. Anything else and it is not.
5. Log in through login.gov
6. Go to https://tdp-backend.app.cloud.gov/admin/
7. You now see the admin console

### This pull request is ready to merge when...

- [x] Meets acceptance criteria for the issue 
- [x] Code is reviewed by someone other than the original author
- [x] Code is tested 
- [x] Experience is approved by UX  
- [x] Meets [a11y checklist](https://github.com/HHS/TANF-app/blob/main/docs/a11y/how-18f-will-test-a11y.md)
- [x] Meets [Raft's Manual QASP Checklist](https://teams.microsoft.com/l/file/B5285301-13CA-46FD-A723-AFFD578C224A?tenantId=d58addea-5053-4a80-8499-ba4d944910df&fileType=docx&objectUrl=https%3A%2F%2Fhhsgov.sharepoint.com%2Fsites%2FTANFDataPortalOFA%2FShared%20Documents%2FGeneral%2FManual%20Testing%20Practices_Issue173.docx&baseUrl=https%3A%2F%2Fhhsgov.sharepoint.com%2Fsites%2FTANFDataPortalOFA&serviceName=teams&threadId=19:f769bbcb029f4f02b55ae7fad90e310d@thread.skype&groupId=41f194a6-c1d3-4680-933e-c8ee7d17e287):lock: (only applicable for PR to main HHS)
- [x] Documentation is updated 
  - OpenAPI  
  - Readme
  - Entity Relationship Diagram (ERD)

### CODE REVIEW

File: docs/Technical-Documentation/user_role_management.md [hunk 1]
```diff
@@ -18,9 +18,17 @@ can can be each be modified through the interface.
...
+- Sign in through `login.gov`
+- Go to `<backend-domain-name>/admin/`
 
-![](images/admin_login.png)
+_Alternatively_
+
+- Go to `<backend-domain-name>/admin/`
+- You will be redirected to `login.gov`
+- Use your `login.gov` credentials to login
+- You will be redirected back to the main application
```
[the remark this answers is not in the document]
Comment (jtwillis92) [reply docs/Technical-Documentation/user_role_management.md:30]: Actually it looks like in `/tdpservice/users/api/utils.py:response_redirect` the redirect is set to always go to the frontend regardless of where they came from. It seems like that redirect back to the admin would be something we'd have to add code for in the TokenAuthorizationOIDC view, but I'm not sure how critical that is to getting this out there.

**Review by jtwillis92 — APPROVED** [review]
I pulled this down locally with the following results:
* Navigated to /admin/ as an unauthenticated user - confirmed redirect to login.gov (note, MFA wasn't prompted but I'm probably still within the period it isn't required again)
* Authenticated through login.gov - confirmed redirected back the the frontend app as expected
* Navigated to /admin/ as an authenticated user without staff flags, confirmed that the admin panel did not give me access and indicated that my user didn't have the permission.
* Navigated to /admin/ as an authenticated user with the staff and superuser flags - confirmed I can now see the admin site and it works as expected

File: docs/Technical-Documentation/user_role_management.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (carltonsmith) [reply docs/Technical-Documentation/user_role_management.md:30]: I don't think it's critical

**Comment by carltonsmith** [comment]
@alexsoble This is deployed and ready to test

**Comment by carltonsmith** [comment]
@ADPennington @alexsoble @lfrohlich This is now deployed to the dev environment and ready for testing

**Comment by alexsoble** [comment]
@ADPennington is reviewing this PR and has blocked out some time tomorrow for review! 😄 

@carltonsmith, one ask to help us with review: Would you mind listing the current `staff` and `superuser` statuses of myself, @ADPennington and @lfrohlich on the dev site? That will help us figure out the best way to test this work on dev. It would be great if each of us had different access levels, so we could manually check all the different access level privileges. 

Noting that since we pushed 510/frontend forward first there's a small merge conflict. Also seeing a `.gitignore`/`cors.json` change that I believe is unrelated? [That said, these are much smaller git diff issues than we had before and I believe this problem won't plague us in the same way now that we're on the single repo model! 😄]

**Comment by carltonsmith** [comment]
> @carltonsmith, one ask to help us with review: Would you mind listing the current `staff` and `superuser` statuses of myself, @ADPennington and @lfrohlich on the dev site? That will help us figure out the best way to test this work on dev. It would be great if each of us had different access levels, so we could manually check all the different access level privileges.

@alexsoble You and @lfrohlich already have the `staff` and `superuser` permissions. @ADPennington has not logged in to the development environment yet, so as of now she does not. Once she has logged in, let me know and I can give her the permissions as well.

**Comment by carltonsmith** [comment]
> Noting that since we pushed 510/frontend forward first there's a small merge conflict. Also seeing a `.gitignore`/`cors.json` change that I believe is unrelated? [That said, these are much smaller git diff issues than we had before and I believe this problem won't plague us in the same way now that we're on the single repo model! 😄]

I have resolved the conflict, so that's taken care of. It's true that the `cors.json` add to `.gitignore` is unrelated. The file was generated as part of the work for 429, so I'll move the change to that PR.

**Comment by ADPennington** [comment]
@carltonsmith i'll need your help with logging in (not sure what credentials to use w/ django admin login). 

one other question before jumping in w/ testing, it'd be helpful to have a bit of clarity on who we have in mind for a `staff` user. Someone like Yun and Thomas? I was having a little trouble thinking about why this kind of user should be able to open the admin console but not be able to read/update in the console.

**Comment by carltonsmith** [comment]
> @carltonsmith i'll need your help with logging in (not sure what credentials to use w/ django admin login).
> 
> one other question before jumping in w/ testing, it'd be helpful to have a bit of clarity on who we have in mind for a `staff` user. Someone like Yun and Thomas? I was having a little trouble thinking about why this kind of user should be able to open the admin console but not be able to read/update in the console.

@ADPennington If this PR is doing what it's supposed to do, you won't need a separate login for the Django Admin. You will be authenticated by your login.gov sign in. So, if you follow the How To Test section above, you should be able to follow along.

As far as `staff` users are concerned, what that flag gives a user is access to the Django Admin. Though, all it is is access. Staff users don't have permissions to read or update data by default. The `superuser` flag is what gives the user permission to read and update data. So, the only users you will want to have these flags set to `True` for are the users who will use Django Admin to read and update data.

**Review by ADPennington — APPROVED** [review]
### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x ] Check against the criteria:
  - [x] Django Admin page protected by two factor authentication
  - [x] Leverage login.gov MFA for authentication to the admin site
  - [x] Only users with staff can open the admin console
  - [x] Only users with superuser can read/update data in the admin console
  - [x] The Admin login page redirects to login.gov

As Product Owner, @lfrohlich will decide if ACs are met.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?

+ Are code coverage minimums met?
  + [x] Backend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [x] Are backend code style checks passing on CircleCI?


### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

+ [ ] Did automated and manual testing with @iamjolly and @ttran-hub using Accessibility Insights reveal any errors introduced in this PR?
_n/a_
    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/a11y/how-18f-will-test-a11y.md)

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

+ [x] If this PR introduces backend code, is that code documented both inline and overall?

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?

no security issues but noticed a weird error page appear when piv authenticated. maybe worth investigating.

**Pull request merged**